### ISSUE TITLE
Question: Support for server sent events?

### ISSUE THREAD

**Issue by whawker** [robtaussig/react-use-websocket#77]
Hi, I was curious if this library supports [server-sents events](https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events)?

I've had a look through the code, and I can't see anything obvious, but I was wondering, as that linked MDN page says the EventSource API is almost identical to the websockets API whether this use case had been raised before?

Great work on this project! I'm especially keen on the concept of sharing a connection so I'm keen on implementing it if I can!

**Comment by robtaussig** [comment]
Looks interesting! It does appear almost identical to the WebSocket API -- I'll look into later and perhaps make a beta release with EventSource support. Thanks for bringing this to my attention!

**Comment by robtaussig** [comment]
Hi @whawker,

Just published d5c3035 to [2.6.0-alpha.0](https://www.npmjs.com/package/react-use-websocket/v/2.6.0-alpha.0) that supports this, and updated the [documentation](https://github.com/robtaussig/react-use-websocket/tree/event_source#eventsourceoptions). Tested it, and it worked with the following:

```js
// Server
var http = require("http");
var fs = require("fs");

http.createServer(function (req, res) {
  var index = "./sse.htm";
  var fileName;
  var interval;
    console.log(req);
  if (req.url === "/")
    fileName = index;
  else
    fileName = "." + req.url;

    console.log(req.url, fileName);

  if (fileName === "./stream") {
    res.writeHead(200, {"Content-Type":"text/event-stream", "Cache-Control":"no-cache", "Connection":"keep-alive", "access-control-allow-origin": "*"});
    res.write("retry: 10000\n");
    res.write("event: connecttime\n");
    res.write("data: " + (new Date()) + "\n\n");
    res.write("data: " + (new Date()) + "\n\n");

    interval = setInterval(function() {
      res.write("data: " + (new Date()) + "\n\n");
    }, 1000);
    req.connection.addListener("close", function () {
      clearInterval(interval);
    }, false);
  } else if (fileName === index) {
    fs.exists(fileName, function(exists) {
      if (exists) {
        fs.readFile(fileName, function(error, content) {
          if (error) {
            res.writeHead(500);
            res.end();
          } else {
            res.writeHead(200, {"Content-Type":"text/html"});
            res.end(content, "utf-8");
          }
        });
      } else {
        res.writeHead(404);
        res.end();
      }
    });
  } else {
    res.writeHead(404);
    res.end();
  }

}).listen(8000, "localhost");
console.log("Server running at http://localhost:8000/");
```

```jsx
// In react component
const {
    lastMessage,
    sendMessage,
    readyState,
  } = useWebSocket('http://localhost:8000/stream', {
    eventSourceOptions: { withCredentials: false }
});
```

Open to any suggestions!

**Comment by whawker** [comment]
Fantastic! Thank you! I will give it a go and get back to you.

In terms of suggestions, I wonder if this should be exposed a separate hook entirely, say `useEventSource`? 

I think they fulfil different use cases, and I think it may cause confusion due to the documented differences? (P.s. thanks for the docs too!)

**Comment by robtaussig** [comment]
Np!

I agree with you and prefer your approach to make it a separate hook. Will implement that before merging with official branch. Thanks!

**Comment by whawker** [comment]
As server sent events are a HTTP protocol connection, rather than a websocket, it would be useful if I could send additional HTTP headers such as `Authorization`?

Would that at all be possible, or does that bloat the API here considerably?

**Comment by whawker** [comment]
It appears I have found out why Event Source is not widely used - there is no way to pass HTTP headers on the request, so I can't use a JWT bearer token to authenticate my request as I assumed. There is the `withCredentials` option to pass Cookies, but this is far from ideal.

I'm not sure what you would like to do with this issue, but it seems likely I'm going to have to explore other technologies such as the [Streams API](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API). I'm incredibly sorry for wasting your time 😖 , thank you so much for helping me on this though, it is much appreciated.

**Comment by whawker** [comment]
We have decided to press ahead with this, and just to avoid sending any sensitive information down the stream. In this sense `EventSource` events act as triggers to fetch data via standard API requests.

Just an observation, [binding directly to the `onmessage` handler](https://github.com/robtaussig/react-use-websocket/blob/event_source/src/lib/attach-shared-listeners.ts#L12), means only events with type `message` are detected, [MDN examples](https://developer.mozilla.org/en-US/docs/Web/API/EventSource#examples) indicate `addEventListener` should be used to detect other message types. 

Do you suppose we could add handlers for different event types via this API? 

Possible API?
```js
const handleNoticeMessage = useCallback((msg) => {
  console.info('Notice', msg)
}, [])

const handleUpdateMessage = useCallback((msg) => {
  console.log('Update', msg)
}, [])

const {
  lastMessage // Maybe this contains last event of any type?
} = useEventSource('http://localhost:8000', {
  withCredentials: false,
  events: {
    notice: handleNoticeMessage,
    update: handleUpdateMessage
  }
})
```

If not, a workaround exists - using `getWebSocket` to obtain the `EventSource` instance, and adding the event handlers manually.

**Comment by robtaussig** [comment]
Hi @whawker,

All great points, and regarding your initial apology about (potentially) wasting my time, that is not necessary at all -- always exciting to learn new web APIs.

That said, I really like your proposed api. A few thoughts:
1) What do you think about adding a `lastEvent` return property that returns any event not covered by `lastMessage` (which would return any event with event type 'message' or without an event type)? I'm not sure how I feel about this, but just throwing it out there.
2) Regarding the `events` property in the `useEventSource` options object, I assume this should be an open-ended type that allows the user to define any events, instead of just an enumerated list? It appears that there is nothing special about `notice` and `update` outside of being popular choices, correct? Are there any issues that you can think of that would be opened up by allowing for dynamically-defined event listeners (since the library would essentially be iterating over the keys and subscribing to each event)?

Finally, in response to your observation that you cannot send headers via WebSockets, that does cause problems for certain use-cases, authentication being the biggest one, but is also a reason why some might be attracted to WebSockets -- information transfer is a lot leaner when you can send messages without the bloat of headers (assuming you don't need them).

**Comment by whawker** [comment]
@robtaussig thanks, apologies about being alarmist originally, it was just frustration of discovering the authentication problem right when we thought everything should slot into place.

> 1. What do you think about adding a `lastEvent` return property that returns any event not covered by `lastMessage` 

Personally I think `lastEvent` should return events of all types, including those of `message` type, I think it might be odd to work out whether `lastMessage` or `lastEvent` has changed, unless you're imagining a pattern with separate`useEffect` hooks for each?

> 2. Regarding the `events` property in the `useEventSource` options object, I assume this should be an open-ended type that allows the user to define any events, instead of just an enumerated list?

Yes sorry, nothing special about `notice` or `update` I imagined any key here, I just used these to easily relate to the MDN example.

> Are there any issues that you can think of that would be opened up by allowing for dynamically-defined event listeners (since the library would essentially be iterating over the keys and subscribing to each event)?

I suppose adding multiple handlers for a single event type could be more confusing with this pattern?

Happy to use the `addEventListener` workaround if you think this would be too restrictive.

**Comment by robtaussig** [comment]
Hi @whawker,

No apology needed. Ok, I'll go with lastEvent including all events, regardless of type. Hopefully should have this out before EOW!

**Comment by whawker** [comment]
Hi @robtaussig can I provide any assistance for the new release?

**Comment by robtaussig** [comment]
@whawker

Apologies for the late reply -- got a little busy at work. I appreciate your offer, but should be able to get this out shortly!

**Comment by robtaussig** [comment]
Hi @whawker,

I merged to master and published to [2.6.0](https://www.npmjs.com/package/react-use-websocket/v/2.6.0).

3fe9ae

Updated README:
https://github.com/robtaussig/react-use-websocket#useeventsource

Let me know if I missed anything!

**Comment by whawker** [comment]
Amazing! Thank you so much! I'll give it a go!
